tensorflow_addons/seq2seq/sampler.py

```diff
@@ -219,14 +223,36 @@ def initialize(self, inputs, sequence_length=None):
             inputs = tf.nest.map_structure(_transpose_batch_time, inputs)
 
         self.input_tas = tf.nest.map_structure(_unstack_ta, inputs)
-        if sequence_length is None:
-            raise ValueError("sequence_length is required for TrainingSampler")
-        self.sequence_length = tf.convert_to_tensor(
-            sequence_length, name="sequence_length")
-        if self.sequence_length.get_shape().ndims != 1:
-            raise ValueError(
-                "Expected sequence_length to be vector, but received shape: %s"
-                % self.sequence_length.get_shape())
+        if sequence_length is not None and mask is not None:
+            raise ValueError("sequence_length and mask can't be provided "
+                             "at the same time.")
+        if sequence_length is not None:
+            self.sequence_length = tf.convert_to_tensor(
+                sequence_length, name="sequence_length")
+            if self.sequence_length.get_shape().ndims != 1:
+                raise ValueError(
+                    "Expected sequence_length to be vector, but received "
+                    "shape: %s" % self.sequence_length.get_shape())
+        elif mask is not None:
+            mask = tf.convert_to_tensor(mask)
+            if mask.get_shape().ndims != 2:
+                raise ValueError(
+                    "Expected mask to a 2D tensor, but received shape: %s" %
+                    mask)
+            if not mask.dtype.is_bool:
+                raise ValueError(
+                    "Expected mask to be a boolean tensor, but received "
+                    "dtype: %s" % repr(mask.dtype))
+
+            axis = 1 if not self.time_major else 0
+            with tf.control_dependencies(
+                [_check_sequence_is_right_padded(mask, self.time_major)]  # pylint: disable=bad-continuation
+            ):
+                self.sequence_length = tf.math.reduce_sum(
+                    tf.cast(mask, tf.int32), axis=axis, name="sequence_length")
+        else:
+            raise ValueError("At least, one of the sequence_length or mask "
```

Comment:
I don't think this is needed.

Reply:
So In the case that none of them is provided, we should assume that the sequence_lengh is a `[max_length]*batch_size` vector?

Reply:
Correct. If neither of them are provided, then it means all the inputs are in full length (we should process all of them).

Reply:
Thanks.

Reply:
Thanks.
One last question, do you have any suggestions on how to infer batch_size from a structured input in a clean way?

I think, First, we need to flatten it (using `tf.nest.flatten`) and then use the first tensor as a reference to infer the batch_size

Reply:
Yea, tf.nest.flatten() and get the first element should be fine.